src/main/java/org/candlepin/auth/CloudRegistrationAuth.java

```diff
@@ -267,23 +275,142 @@ public String generateRegistrationToken(Principal principal, CloudRegistrationIn
+        }
+
+        if (principal == null) {
+            throw new IllegalArgumentException("principal is null");
+        }
+
+        if (cloudRegInfo == null) {
+            throw new IllegalArgumentException("cloudRegistrationInfo is null");
+        }
+
+        CloudAuthenticationResult result = this.cloudRegistrationAdapter
+            .resolveCloudRegistrationDataV2(cloudRegInfo);
+
+        String cloudAccountId = result.getCloudAccountId();
+        if (cloudAccountId == null || cloudAccountId.isBlank()) {
+            String errmsg = this.i18nProvider.get()
+                .tr("cloud provider account could not be resolved");
+
+            throw new CloudRegistrationAuthorizationException(errmsg);
+        }
+
+        String cloudInstanceId = result.getCloudInstanceId();
+        if (cloudInstanceId == null || cloudInstanceId.isBlank()) {
+            String errmsg = this.i18nProvider.get()
+                .tr("cloud instance ID could not be resolved");
+
+            throw new CloudRegistrationAuthorizationException(errmsg);
+        }
+
+        String cloudProviderShortName = result.getCloudProviderShortName();
+        if (cloudProviderShortName == null || cloudProviderShortName.isBlank()) {
+            String errmsg = this.i18nProvider.get()
+                .tr("cloud provider short name could not be resolved");
+
+            throw new CloudRegistrationAuthorizationException(errmsg);
+        }
+
+        String productId = result.getProductId();
+        if (productId == null || productId.isBlank()) {
+            String errmsg = this.i18nProvider.get()
+                .tr("product id could not be resolved");
+
+            throw new CloudRegistrationAuthorizationException(errmsg);
+        }
+
+        String ownerKey = result.getOwnerKey();
+        boolean ownerAndPoolExists = false;
+        if (ownerKey != null && !ownerKey.isBlank() && result.isEntitled()) {
+            Owner owner = ownerCurator.getByKey(ownerKey);
+            if (owner != null) {
+                ownerAndPoolExists = !poolCurator.listByOwnerAndProduct(owner, productId).isEmpty();
+            }
+        }
+
+        String anonymousConsumerUuid = null;
+        if (ownerKey == null || ownerKey.isBlank() || !result.isEntitled() || !ownerAndPoolExists) {
+
+            // TODO: Schedule anonymous org creation job
+
+            AnonymousCloudConsumer anonConsumer = anonymousCloudConsumerCurator
+                .getByCloudInstanceId(cloudInstanceId);
+
+            if (anonConsumer == null) {
+                anonConsumer = new AnonymousCloudConsumer()
+                    .setCloudAccountId(cloudAccountId)
+                    .setCloudInstanceId(cloudInstanceId)
+                    .setProductId(productId)
+                    .setCloudProviderShortName(cloudProviderShortName);
+
+                anonConsumer = anonymousCloudConsumerCurator.create(anonConsumer);
```

Comment:
Something we need to think about here is where should the transactional boundary be. We aren't within a Resource method (all of them which have @transactional set on). This is the first time that we actually persist new data in an authentication method. 

What happens if we create the consumer here, but we fail later on (e.g. when we generate the token)? Probably we need a @Transactional on the `generateRegistrationTokenV2` method itself

Reply:
Given how the base design this is being built on works, it *should* be in the context of a resource method. The token generation is done as a specific request from a method which currently doesn't do any transaction management, but could easily enough.

The more I look at this and the other comment regarding the splitting up of classes, I wonder if this needs to be broken up even further into not just two auth classes as suggested in the other comment, but isolating the auth bits into *only* the token verification, and moving token generation to the respective resource or some other shared utility. Thoughts on this?

Reply:
I got confused for a moment because we're in an *Auth.java method, but you're right. We are in the context of the /cloud/authorize resource method, so that's where the transactional annotation should be set.

And me being confused by this, is proof that we should be splitting this class even further. So something like:
CloudRegistrationAuthV1: handles only the v1 autoreg verification
CloudRegistrationAuthV2: handles only the v2 autoreg verification
Some other class that deals with generating the v1 token, and is called by the Resource method
Some other class that deals with generating the v2 token, and is called by the Resource method

Reply:
I agree with having two cloud registration auth classes (CloudRegistrationAuth and AnonymousCloudRegistrationAuth) that only handles verifying the http request authorization handler and producing a Principal as that is their responsibility as a AuthProvider. Since this authorization is done differently between the two token types, it is nice to separate that logic into two different classes instead of having one class with bunch of conditionals based on token type (as Chris mentioned).

It is a little strange to be using an AuthProvider class for token generation and so I agree that moving the token generation logic something like a utility method or maybe to CloudRegistrationResource makes sense to me.

Reply:
I added a commit showing the restructuring of the cloud registration auth classes and the token generation. I moved the `CloudRegistrationAdapter` calls and the processing of the result to the `CloudRegistrationResource` class. I want to check and make sure this aligns with the separation that you had in mind!

Reply:
The general separation looks good. For more detailed feedback I'll add new comments